implement pod exec websockets v5

aojea opened the pull request:
/kind feature
/kind api-change
/kind deprecation

```release-note
support new exec v5 websocket subprotocol
```

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets#proposal-new-remotecommand-sub-protocol-version---v5channelk8sio
```

cc: @yliaog @siyuanfoundation

aojea committed:
implement pod exec v5

implementation is able to signal a channel is closed to the other side.
Clients are also able to drain the closed channels.

seans3 commented:
/remove-lifecycle rotten

seans3 commented:
/assign

seans3 commented:
/lifecycle frozen

seans3 commented on kubernetes/base/stream/ws_client.py:
```diff
@@ -104,6 +111,14 @@ def readline_channel(self, channel, timeout=None):
                         del self._channels[channel]
                     return ret
+
+            if channel in self._closed_channels:
```
Do we need this same short-circuiting code in `peek_channel` and `read_channel` ?

aojea replied:
ack

seans3 commented on kubernetes/base/stream/ws_client.py:
```diff
@@ -119,6 +134,14 @@ def write_channel(self, channel, data):
+        """Close a channel (v5 protocol only)."""
+        if self.subprotocol != V5_CHANNEL_PROTOCOL:
+            return
+        data = bytes([255, channel])
```
Should we define a constant (e.g. CLOSE_CHANNEL = 255 or V5_HALF_CLOSE = 255) for this magic number, and reference the constant?

yliaog replied:
+1 to have a constant

aojea replied:
ack

seans3 commented on kubernetes/base/stream/ws_client_test.py:
```diff
@@ -123,6 +126,117 @@ def test_websocket_proxycare(self):
+
+            mock_ws.send.assert_not_called()
+
+    def test_update_receives_close_v5(self):
```
Should we add an additional unit test to verify how `readline_channel` handles a closed channel with leftover data?

While the current unit tests cover the parsing of the close signal itself, the new logic you added inside `readline_channel`—which flushes the remaining buffer even if it lacks a newline—is currently untested. Having a test that asserts `readline_channel` successfully flushes leftover data (e.g. "hello") and then returns an empty string on the subsequent call would ensure this specific edge-case logic is protected from future regressions.

aojea replied:
added tests also for `read_channel` and `peek_channel` to validate channels are drained and follow the expected semantics

aojea replied:
added

seans3 commented on kubernetes/e2e_test/test_client.py:
```diff
@@ -201,6 +201,63 @@ def test_pod_apis(self):
+            # Wait for process to exit
+            resp.update(timeout=5)
+            start = time.time()
+            while resp.is_open() and time.time() - start < 10:
```
Should we simplify this polling loop by using resp.run_forever(timeout=15) instead of manually checking resp.is_open()? This would align it closely with how the other exec tests in this file wait for streams to close.

aojea replied:
done

seans3 commented on kubernetes/base/stream/ws_client_test.py:
```diff
@@ -123,6 +126,117 @@ def test_websocket_proxycare(self):
+
+            # Setup frame with close signal for channel 0
+            frame = MagicMock()
+            frame.data = b'\xff\x00'
```
Would this be better as constants, such as `frame.data = bytes([CLOSE_CHANNEL, STDIN_CHANNEL])` or whatever the 255 constant is defined as.  I believe other places that we see `b'\xff\x00'` would be more readable with the constants.

aojea replied:
moved magic numbers to constants

seans3 reviewed (changes requested):
This looks good and substantially complete.

One high-level observation: I believe falling back to `v4.channel.k8s.io` is the right approach for backwards compatibility, but we should probably note that for clusters < v1.30, the client will silently revert to the broken behavior and hang on EOF. 

I've left a few minor inline comments mostly about tests and constants.

seans3 commented:
/assign @yliaog

yliaog commented on kubernetes/base/stream/ws_client.py:
```diff
@@ -200,13 +223,24 @@ def update(self, timeout=0):
+                    # In Py3, iterating bytes gives int, but indexing bytes gives int.
+                    # websocket-client frame.data might be bytes.
+
+                    if channel == 255 and self.subprotocol == V5_CHANNEL_PROTOCOL: # v5 CLOSE
```
better to replace 255 with a constant

aojea replied:
done

aojea commented:
@seans3 @yliaog PTAL

seans3 commented:
Looks great--thanks.

/lgtm

yliaog commented on kubernetes/e2e_test/test_client.py:
```diff
@@ -201,6 +201,60 @@ def test_pod_apis(self):
+        if resp.subprotocol != "v5.channel.k8s.io":
+            resp.close()
+            api.delete_namespaced_pod(name=name, body={}, namespace='default')
+            self.skipTest("Skipping test: v5.channel.k8s.io subprotocol not negotiated")
```
the test is skipped in the e2e test flow: https://github.com/kubernetes-client/python/actions/runs/25740181177/job/75636044165

how to make it run?

yliaog replied:
kubernetes/e2e_test/test_client.py::TestClient::test_pod_exec_close_channel SKIPPED

aojea replied:
we need to get a more recent version of kind, let me update that

aojea replied:
it was not the version, the e2e tests were not initializing the subprotocol so it never matched ... now the client if is not preconfigured, uses the subprotocol obtained during the negotiation

aojea committed:
use latest stable kind version

aojea committed:
use subprotocols from handshake

aojea commented:
@yliaog is it normal for the e2e tests to take so long? they take more than 3 hours, that sounds like a lot

aojea committed:
fix(watch): shutdown raw socket in stop() to prevent SSL read deadlocks

aojea commented:
> 6a83c7e

it runs in less than 2 minutes in my vm https://github.com/kubernetes-client/python/pull/2486/commits/6a83c7e951ecb396e2afeef8d2aedcddf86214c1 , it seems there can be a deadlock on stop

yliaog commented:
the e2e takes a long time in the informer tests, i have not had a chance to look into why it takes so long.

aojea commented:
> the e2e takes a long time in the informer tests, i have not had a chance to look into why it takes so long.

Fixed im the latest commit, there was deadlockimg when.closing, added a comment

yliaog commented:
thanks for the fix @aojea 

thanks for the review @seans3 

/lgtm
/approve
